packages/fx-core/src/core/middleware/projectUpgrader.ts

```diff
@@ -0,0 +1,202 @@
+import { Middleware, NextFunction } from "@feathersjs/hooks";
+import { ConfigFolderName, err, Inputs, Json, ProjectSettings } from "@microsoft/teamsfx-api";
+import {
+  ContextUpgradeError,
+  CoreHookContext,
+  FxCore,
+  NoProjectOpenedError,
+  PathNotExistError,
+} from "..";
+import * as fs from "fs-extra";
+import * as path from "path";
+import { deserializeDict, serializeDict } from "../..";
+
+const resourceContext = [
+  {
+    plugin: "fx-resource-aad-app-for-teams",
+    secret: "local_clientSecret",
+    relatedKeys: [
+      "local_clientId",
+      "local_objectId",
+      "local_oauth2PermissionScopeId",
+      "local_tenantId",
+      "local_applicationIdUris",
+    ],
+  },
+  {
+    plugin: "fx-resource-bot",
+    secret: "localBotPassword",
+    relatedKeys: ["localBotId", "localObjectId", "local_redirectUri", "bots", "composeExtensions"],
+  },
+];
+
+const solutionContext = {
+  plugin: "solution",
+  relatedKeys: ["localDebugTeamsAppId", "teamsAppTenantId"],
+};
+
+export const ProjectUpgraderMW: Middleware = async (ctx: CoreHookContext, next: NextFunction) => {
+  await upgradeContext(ctx);
+  await next();
+};
+
+export async function upgradeContext(ctx: CoreHookContext): Promise<void> {
+  try {
+    const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+    if (!inputs.projectPath) {
+      ctx.result = err(NoProjectOpenedError());
+      return;
+    }
+    const projectPathExist = await fs.pathExists(inputs.projectPath);
```

Comment:
I'm not sure about the design philosophy about middleware, is these checks (whether project is opened or project path exists) already happened globally? If every middleware needs to check this, there would be many duplicate logic...

Reply:
Here I add the logic since this part may be added to some cmds in cli and I can not make sure whether other middlewares will do the check. Maybe @jayzhang can consider to make the check in a separate middleware?